### PR TITLE
✨ Adding `generateCacheKey` to `ApolloServerPluginResponseCache` to allow for custom cache keys

### PR DESCRIPTION
**Pull request by kschrade**
Create a new hook (`generateCacheKey`) on the `ApolloServerPluginResponseCache`
constructor options which allows users to customize the cache key used by returning
their own `string`.

The hook receives two arguments: the `requestContext` and an object (`cacheKeyData`)
which contains the following properties:
`source`
`operationName`
`variables`
`extra` (computed by `options.extraCacheKeyData`)
`sessionId`
`sessionMode`

Landing in favor of:
#6649
#6428

### CODE REVIEW

File: packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts [hunk 1]
```diff
@@ -165,10 +172,15 @@ export default function plugin(
...
             contextualCacheKeyFields: ContextualCacheKey,
           ): Promise<GraphQLResponse | null> {
-            const key = cacheKeyString({
+            const cacheKeyData = {
               ...baseCacheKey!,
               ...contextualCacheKeyFields,
-            });
+            };
+
+            const key = options.generateCacheKey
```
Comment (glasser) [inline packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:180]: we can do this once by, higher up, having something like `const generateCacheKey = options.generateCacheKey ?? (_, key) => sha(JSON.stringify(key))`

File: packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts [hunk 2]
```diff
@@ -86,6 +86,13 @@ interface Options<TContext = Record<string, any>> {
   shouldWriteToCache?(
     requestContext: GraphQLRequestContext<TContext>,
   ): ValueOrPromise<boolean>;
+
+  // This hook allows one to replace the function that is used to create a cache
+  // key. If not it will fall back to what happens today.
```
Comment (glasser) [inline packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:91]:
```suggestion
  // key. By default, it is the SHA-256 (from the Node `crypto` package) of the result of
  // calling `JSON.stringify(keyData)`. You can override this to customize the serialization
  // or the hash, or to make other changes like adding a prefix to keys to allow for
  // app-specific prefix-based cache invalidation. You may assume that `keyData` is an object
  // and that all relevant data will be found by the kind of iteration performed by
  // `JSON.stringify`, but you should not assume anything about the particular fields on
  // `keyData`.
```

How about something like this?

**Review by glasser — CHANGES_REQUESTED** [review]
Please also provide at least a brief mention of this parameter in the docs (`docs/source/performance/caching.md`). The table in the bottom section seems reasonable. Thanks1

File: packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts (hunk 1 above)
Comment (kschrade) [reply packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:186]: done - in next push

File: packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts [hunk 3]
```diff
@@ -155,6 +155,10 @@ export default function plugin(
         'fqc:',
       );
 
+      const generateCacheKey = options.generateCacheKey
+        ? options.generateCacheKey
+        : (_: any, key: any) => sha(JSON.stringify(key));
```
Comment (glasser) [inline packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts:166]: Why `any`? I think we can usually avoid `any`.

Note that you can probably delete the cacheKeyString function now.

File: docs/source/performance/caching.md [hunk 4]
```diff
@@ -472,3 +472,4 @@ In addition to [the `sessionId` function](#identifying-users-for-private-respons
 | `extraCacheKeyData` | This function's return value (any JSON-stringifiable object) is added to the key for the cached response. For example, if your API includes translatable text, this function can return a string derived from `requestContext.request.http.headers.get('Accept-Language')`. |
 | `shouldReadFromCache` | If this function returns `false`, Apollo Server _skips_ the cache for the incoming operation, even if a valid response is available. |
 | `shouldWriteToCache` | If this function returns `false`, Apollo Server doesn't cache its response for the incoming operation, even if the response's `maxAge` is greater than `0`. |
+| `generateCacheKey` | This function's return is used as the key for the cache. If it is not overrode a sha256 hash is used |
```
Comment (glasser) [inline docs/source/performance/caching.md:475]:
```suggestion
| `generateCacheKey` | Customize generation of the cache key. By default, this is the SHA256 hash of the JSON encoding of an object containing relevant data. |
```

something like this? (I don't think "overrode" is a word, and your version was missing the trailing period.)

**Review by glasser — CHANGES_REQUESTED** [review]

File: packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts (hunk 3 above)
Comment (kschrade) [reply packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts]: > Why `any`? I think we can usually avoid `any`.

fair point, I don't know why I left it that way. Updated it now with a type that makes it strongly typed.

> Note that you can probably delete the cacheKeyString function now.

it is gone now.

Comment (kschrade) [reply packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts]: thank your @trevor-scheer for the assist

**Pull request merged**